iothub/device/tests/HsmAuthentication/ModuleAuthenticationWithHsmTest.cs

```diff
@@ -33,16 +33,44 @@ public async Task TestSafeCreateNewToken_ShouldReturnSasToken()
             string sasToken = await moduleAuthenticationWithHsm.GetTokenAsync(this.iotHub);
             SharedAccessSignature token = SharedAccessSignature.Parse(iotHub, sasToken);
 
-            string audience = WebUtility.UrlEncode(string.Format(CultureInfo.InvariantCulture, "{0}/devices/{1}/modules/{2}",
+            string audience = string.Format(CultureInfo.InvariantCulture, "{0}/devices/{1}/modules/{2}",
                 this.iotHub,
-                this.deviceId,
-                this.moduleId));
+                WebUtility.UrlEncode(this.deviceId),
+                WebUtility.UrlEncode(this.moduleId));
 
             // Assert
             httpClient.Verify();
             Assert.IsNotNull(sasToken);
             Assert.AreEqual(this.signature, token.Signature);
-            Assert.AreEqual(WebUtility.UrlDecode(audience), token.Audience);
+            Assert.AreEqual(audience, token.Audience);
+            Assert.AreEqual(string.Empty, token.KeyName);
+        }
+
+        [TestMethod]
+        public async Task TestSafeCreateNewToken_ShouldReturnSasToken_DeviceIdWithChars()
+        {
+            // Arrange
+            string deviceId = "n@m.et#st";
+            string moduleId = "$edgeAgent";
+            var httpClient = new Mock<ISignatureProvider>();
```

Comment:
Is it possible to add E2E for this scenario ?

Reply:
Any Edge E2E scenario with Device IDs that have special characters should cover it. 
We will probably add E2E tests for this as well, once the SDK gets fixed.

Reply:
Can we get an E2E test where we create such a device, use it then delete it? I'm not very familiar with the HSM Module code or how to test this out with Edge or IotHub.